### PR TITLE
mgjson: A miniature version of the Mongoose Library, includes only th…

### PR DESCRIPTION
**Pull request by TheRootED24**
Maintainer: Scott Mercer @TheRootED24
Compile tested: (ATH79, gl-inet, OpenWrt version r27160-b72c4b5386)
Run tested: (ATH79, gl-inet, OpenWrt version r27160-b72c4b5386, tests done Basic Functionality, valgrind 0 errors 0 bytes in use at exit)

Description:
mgjson: A miniature version of the Mongoose Library, includes only the mg_json, mg_str and mg_iobuf functionality of the Mongoose Library, without the overhead of the Networking.

### CODE REVIEW

**Comment by BKPepe** [comment]
Would it be possible to stop creating so many pull requests? In history, it is possible to see that you opened nearly 15 PRs, most of which are closed. There is the possibility to force push it, which should help you. Also, I see that you are trying to add new packages into OpenWrt. The best approach would be for you to do it one by one, because all of the Makefiles you prepared are not exactly right. So, do one properly, once it is in, you can adjust the others. Not big deal. :-)

**Comment by TheRootED24** [comment]
Yes, i apologize. These are my first actual commits ... i asked for help in the forums,and pm'd a couple of devs, got nothing in the terms of help. On the bright-side of things, I'm pretty sure i've got it figured out now :)

I did as you suggest, only i added 2 packages, because one is a dependent library. jsonmg links against mgjson.

Any tips hints suggestions are graciously recieved.

File: libs/mgjson/Makefile
Comment (BKPepe) [inline libs/mgjson/Makefile:14]: I checked it, your repository is only 3 weeks old. 
- While looking at it, I found that commit messages are vague or should I rather say that all of them looks the same?
- No one reviewed it.
- There is no documentation except the README, which does not say anything.

File: libs/mgjson/Makefile
Comment (BKPepe) [inline libs/mgjson/Makefile:13]: This variable is for what? How it is used?

File: libs/mgjson/Makefile
Comment (BKPepe) [inline libs/mgjson/Makefile:6]: You should look in the CONTRIBUTING.md file in this repository, which says:
> Provide an up-to-date Copyright notice or none. Copyright should not be assigned to OpenWrt unless you are explicitly requested by or working under contract to OpenWrt. Assigning a Copyright to yourself or organization you represent is acceptable.

File: libs/mgjson/Makefile
Comment (BKPepe) [inline libs/mgjson/Makefile:32]: How did you tested it? Have you checked ``make menuconfig``?

File: libs/mgjson/Makefile
Comment (BKPepe) [inline libs/mgjson/Makefile:35]: Do we need to have ``-stdu=gnu99``?

File: libs/mgjson/Makefile
Comment (BKPepe) [inline libs/mgjson/Makefile:48]: Why do you need to have this?

**Review by BKPepe — CHANGES_REQUESTED** [review]

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:6]: again, a question i asked in  many places in which  you'd expect a thorough explanation ... got nothing. I assumed the copyright was some how tied to the example makefile i used originally, provided by the OpenWrt build system. 

Is there a better resource to have these types of questions answered ?

As i said in my previous comment ... these are my first commits. I have done what i consider my due diligence.  I have read any and all documentation offered.  I have no problem making any changes you suggest. I have tested everything to the absolute best of my abilities. I have compile with the all the relevant flags that I know of for error checking and have no errors anywhere in compilation , gcc is happy, valgrind reports no errors no leaks. I have tested everything thoroughly inside of the openwrt build environment, (yes make menuconfig loads the packages with no complaints, again everything is as expected.  

I will make the changes np, in fact I welcome the review, as  I have been requesting a code review in the forums to try avoid this exact situation. Please just let me know what you want changed/fixed, and I will gladly make the changes. If you can find any issue what so ever with my sources, please do point them out and i will be happy to investigate. Please just don't talk down to me like I should know all of this ...when no one seems to have time to answer these questions where they should in fact be answered.

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:32]: yes, all build and basic functionality have been tested thoroughly. 

Screenshot from 2024-08-21 13-05-00
Screenshot from 2024-08-21 13-03-35

I never included a description for the mgjson library, this should probably be included ?

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:35]: no, that is no longer required.

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:48]: As i said in my original comment, this is just a shared library, used by the second package i submitted (jsonmg) to link against, therefore the includes from mgjson are required(mgjson.h), and the libmgjson.so in required at runtime.

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:13]: It was in the sample i started with. I wasn't sure of its significance, and again, could not get an answer as to whether or not it was required

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:48]: I would like to give the user the option to compile static, or dynamically. given the size, static is likely the best option,  I thought there was some way i could offer this in the menuconfig ...  however i couldn't find it in the docs and again ... no answer in the forums

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:14]: yes, this is fair. This library was a after the fact addition. I made it after it was brought to my attention that the jsonmg library had far more utility as a standalone library, rather then being strapped to the rather large mongoose library. Being a shared library, not to mention, the library itself is part of the very well documented Mongoose C Networking library -->[https://mongoose.ws/documentation/](url) i guess i didn't figure separate documentation was needed at the time. I do however, see your point, i'll add some more general info to my repo with more explaination regarding the library's

File: libs/mgjson/Makefile
Comment (BKPepe) [reply libs/mgjson/Makefile:32]: You shared two screenshots. The first one even shows that something is wrong. Title is not shown as it is too long.

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:48]: i just was pointing here --> [https://openwrt.org/docs/guide-developer/frequent-pr-mistakes](url)

This would've been of great help.I see allot of mistakes could've been avoided.  My question now is. What is expected of me now, do I make the edit/s and submit a new pr, do you make the edits ? 

I agree with all proposed changes ...other than, the necessity of the inclusion of the #define Build/dev directives, which unless I am mistaken, is a valid inclusion ??

Comment (TheRootED24) [reply libs/mgjson/Makefile:48]: In regards to adding more general info to the mgjson repo. Is this something you want done during review. I just do not what to add unnecessary commits during review. Please advise.

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:32]: yes, I mentioned that in the comment attached to the photos. This was an oversight on my part, I forgot to add a description field. This library, as you mentioned, is only 3 weeks from conception. I probably should've have started with the more mature library first. In any case. i will add a the description field.

Comment (TheRootED24) [reply libs/mgjson/Makefile:32]: If the title is to long, I am fine trimming it to the period ie.
```
A Stripped down OpenWrt version of Mongoose C Networking.
```

**Comment by TheRootED24** [comment]
I have address all issues found. However, The jsonmg package fails to link without the inclusion of the define Build/InstallDev
```

define Build/InstallDev
	$(INSTALL_DIR) $(1)/usr/include/mgjson
	$(CP) \
		$(PKG_BUILD_DIR)/src/*.h \
		$(1)/usr/include/mgjson/

	$(INSTALL_DIR) $(1)/usr/lib
	$(CP) \
		$(PKG_BUILD_DIR)/src/libmgjson.so \
		$(1)/usr/lib/

endef

```
I also tried just including the header  file without the .so in the staging directory, and it fails to link. 

I have removed the cr, fixed the title, to which i've included screen shots, demonstrating the fixed title and the automatic inclusion triggered by the selection of the dependent jsonmg pkg

And I have included the build log from today of the jsonmg package, which builds and links to the mgjson package successfully, no errors that I could spot, tho my eyes are getting tired of looking at it :).
[jsonmg_build_log_08-21-24.zip](https://github.com/user-attachments/files/16701052/jsonmg_build_log_08-21-24.zip)

Still waiting for approval to push amended Makefile.  And what you would like in regards to updating the general info of the mgjson repo. Is it priority or can I do it after package has been accepted. Please advise

Also included the amended Makefile
[Makefile.zip](https://github.com/user-attachments/files/16701068/Makefile.zip)

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:35]: I have removed line (35) entirely in the amended file.

File: libs/mgjson/Makefile
Comment (TheRootED24) [reply libs/mgjson/Makefile:6]: the copyright has been removed, as per your request

**Comment by TheRootED24** [comment]
If there is anything else I can do, please let me know. I have 3 more packages to add yet.  I need to get them off my desk before I can get started on support for the ipq50XX devices.

**Comment by TheRootED24** [comment]
Should i close this pr and send another ? Please, any help here is greatly appreciated ?

**Comment by TheRootED24** [comment]
Thanks for your help.  I see I have more reading to do. Sorry to have wasted your time

**Pull request closed without merging**
